### PR TITLE
[#3751] feat(CI): Enable Apache Ranger plugin in Hive Docker image

### PR DESCRIPTION
**Pull request by unknowntpo**
### What changes were proposed in this pull request?

1. Enable Apache Ranger plugin in Hive Docker image.
2. Add Ranger integration Hive permission test.
3. Both support Hive2 and Hive3
   * To start the Hive container, you can to specify environment variables: HIVE_RUNTIME_VERSION:
   * Hive version, currently support `hive2`(default) and `hive3` DOCKER_ENV_RANGER_SERVER_URL:
   * Ranger server URL DOCKER_ENV_RANGER_HIVE_REPOSITORY_NAME: Ranger Hive repository name
   * DOCKER_ENV_RANGER_HDFS_REPOSITORY_NAME: Ranger HDFS repository name

Fix: #3751

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Local test.

TODO:

- [x] @xunliu needs to publish hive docker image
- [x] @unknowntpo needs to update docker-image-details.md

### CODE REVIEW

**Comment by xunliu** [comment]
hi @unknowntpo Thank you for your contributions.
Please help build two hive dockers to test Gravitino integration test.
1. HADOOP_VERSION=2.7.3 HIVE_VERSION=2.3.9 ./dev/docker/build-docker.sh --platform linux/arm64 --type hive --image datastrato/hive --tag 2.3.9-ranger
2. HADOOP_VERSION=3.1.0 HIVE_VERSION=3.1.3 ./dev/docker/build-docker.sh --platform linux/arm64 --type hive --image datastrato/hive --tag 3.1.3-ranger

**Comment by unknowntpo** [comment]
I've published [docker images](https://hub.docker.com/r/unknowntpo/hive) with these tags.

These two commands passed these tests:

- CatalogHiveIT
- CatalogIcebergRestIT
- ProxyCatalogHiveIT

Note that if we run these tests with Intellij, we got `not tests found for given include xxxx error`,
in order to run tests, we needs to modify tag from:

```
@Tag("gravitino-docker-test")
```

to

```
@Tag("gravitino-docker-it")
```

Screenshot 2024-06-29 at 4 57 54 PM

File: dev/docker/ranger/init-mysql.sql.template
[the remark this answers is not in the document]
Comment (unknowntpo) [reply dev/docker/ranger/init-mysql.sql.template:1]: ok, reverted.

File: dev/docker/ranger/start-ranger-services.sh
[the remark this answers is not in the document]
Comment (unknowntpo) [reply dev/docker/ranger/start-ranger-services.sh:1]: ok, reverted.

File: dev/docker/hive/hive-dependency.sh [hunk 1]
```diff
@@ -8,19 +8,30 @@ hive_dir="$(dirname "${BASH_SOURCE-$0}")"
...
 JDBC_DIVER_DOWNLOAD_URL="https://downloads.mysql.com/archives/get/p/3/file/${JDBC_DIVER_PACKAGE_NAME}"
 
+ZOOKEEPER_PACKAGE_NAME="zookeeper-${ZOOKEEPER_VERSION}.tar.gz"
+ZOOKEEPER_DOWNLOAD_URL="https://archive.apache.org/dist/zookeeper/zookeeper-${ZOOKEEPER_VERSION}/${ZOOKEEPER_PACKAGE_NAME}"
+
+RANGER_HIVE_PACKAGE_NAME="ranger-distro-${RANGER_VERSION}-hive-plugin.tar.gz"
+RANGER_HIVE_DOWNLOAD_URL="https://repo.maven.apache.org/maven2/org/apache/ranger/ranger-distro/${RANGER_VERSION}/${RANGER_HIVE_PACKAGE_NAME}"
+
+RANGER_HDFS_PACKAGE_NAME="ranger-distro-${RANGER_VERSION}-hdfs-plugin.tar.gz"
+RANGER_HDFS_DOWNLOAD_URL="https://repo.maven.apache.org/maven2/org/apache/ranger/ranger-distro/${RANGER_VERSION}/${RANGER_HDFS_PACKAGE_NAME}"
```
[the remark this answers is not in the document]
Comment (unknowntpo) [reply dev/docker/hive/hive-dependency.sh]: done.

**Comment by xunliu** [comment]
> I've published [docker images](https://hub.docker.com/r/unknowntpo/hive) with these tags.
> 
> These two commands passed these tests:
> 
> * CatalogHiveIT
> * CatalogIcebergRestIT
> * ProxyCatalogHiveIT
> 
> Note that if we run these tests with Intellij, we got `not tests found for given include xxxx error`, in order to run tests, we needs to modify tag from:
> 
> ```
> @Tag("gravitino-docker-test")
> ```
> 
> to
> 
> ```
> @Tag("gravitino-docker-it")
> ```
> 
> Screenshot 2024-06-29 at 4 57 54 PM

hi @unknowntpo 
Currently Gravitino didn't automatic run Docker integration test.
If you need run Docker IT in your local, please set `skipDockerTests = false` in 
 the [gradle.properties:L29](https://github.com/datastrato/gravitino/blob/main/gradle.properties#L29)

So, we didn't need to modify `@Tag("gravitino-docker-test")`.

File: dev/docker/hive/core-site.xml
Comment (xunliu) [inline dev/docker/hive/core-site.xml:1]: Please add these codes in the core-site.xml, Because we need a user(`datastrato`) to test proxy other use to operation hive.

```
  <property>
    <name>hadoop.proxyuser.datastrato.groups</name>
    <value>*</value>
  </property>

  <property>
    <name>hadoop.proxyuser.datastrato.hosts</name>
    <value>*</value>
  </property>
```

File: dev/docker/hive/Dockerfile [hunk 2]
```diff
@@ -83,15 +86,31 @@ RUN echo "HADOOP_CONF_DIR=${HADOOP_CONF_DIR}" >> /etc/environment
...
 
+################################################################################
+# install zoofkeeper
+ADD packages/zookeeper-${ZOOKEEPER_VERSION}.tar.gz /opt/
+RUN ln -s /opt/zookeeper-${ZOOKEEPER_VERSION} ${ZK_HOME}
+RUN mkdir $ZK_HOME/data
+
+################################################################################
+# install ranger hadoop plugin
+ADD packages/ranger-distro-${RANGER_VERSION}-hdfs-plugin.tar.gz /opt/
```
Comment (xunliu) [inline dev/docker/hive/Dockerfile:102]: I think `ranger-distro-${RANGER_VERSION}-hdfs-plugin.tar.gz` is incorrect, It should be `ranger-${RANGER_VERSION}-hdfs-plugin.tar.gz`

File: dev/docker/hive/Dockerfile [hunk 3]
```diff
@@ -83,15 +86,31 @@ RUN echo "HADOOP_CONF_DIR=${HADOOP_CONF_DIR}" >> /etc/environment
...
+RUN ln -s /opt/ranger-${RANGER_VERSION}-hdfs-plugin ${RANGER_HDFS_PLUGIN_HOME}
+
+################################################################################
+# install ranger hive plugin
+ADD packages/ranger-distro-${RANGER_VERSION}-hive-plugin.tar.gz /opt/
```
Comment (xunliu) [inline dev/docker/hive/Dockerfile:107]: I think `ranger-distro-${RANGER_VERSION}-hive-plugin.tar.gz` is incorrect, It should be `ranger-${RANGER_VERSION}-hive-plugin.tar.gz`?

File: dev/docker/hive/Dockerfile (hunk 2 above)
Comment (unknowntpo) [reply dev/docker/hive/Dockerfile]: ok, fixed.

File: dev/docker/hive/Dockerfile (hunk 3 above)
Comment (unknowntpo) [reply dev/docker/hive/Dockerfile:107]: ok, fixed

**Comment by xunliu** [comment]
@mchades @yuqi1129 @jerryshao Please help me review this PR, Thanks.

File: dev/docker/hive/Dockerfile [hunk 4]
```diff
@@ -105,14 +124,10 @@ ADD hdfs-site.xml ${HADOOP_CONF_DIR}/hdfs-site.xml
 ADD mapred-site.xml ${HADOOP_CONF_DIR}/mapred-site.xml
 ADD check-status.sh /tmp/check-status.sh
 
-# format HFS
-RUN ${HADOOP_HOME}/bin/hdfs namenode -format -nonInteractive
```
Comment (yuqi1129) [inline dev/docker/hive/Dockerfile:109]: Why do we remove this?

File: integration-test/build.gradle.kts [hunk 5]
```diff
@@ -108,6 +108,10 @@ dependencies {
   testImplementation(libs.trino.client) {
     exclude("jakarta.annotation")
   }
+  testImplementation(libs.hive2.jdbc) {
+    4
```
Comment (yuqi1129) [inline integration-test/build.gradle.kts:112]: What does "4" mean here?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+@Tag("gravitino-docker-it")
+public class RangerHiveIT extends RangerIT {
+  private static final ContainerSuite containerSuite = ContainerSuite.getInstance();
+  private static Connection adminConnection;
+  private static Connection anonymousConnection;
+  private static final String adminUser = "datastrato";
+  private static final String anonymouslUser = "anonymous";
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java:33]: anonymouslUser typo

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java [hunk 7]
```diff
@@ -5,69 +5,313 @@
...
+
+      if (cleanAllPolicy) {
+        cleanAllPolicy(RANGER_HDFS_REPO_NAME);
+      }
+    } catch (RangerServiceException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected void createRangerHivePolicy(
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java:205]: This method has not been utilized.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java [hunk 8]
```diff
@@ -5,69 +5,313 @@
...
+            .build());
+
+    try {
+      RangerService createdService = rangerClient.createService(service);
+      Assertions.assertNotNull(createdService);
+
+      Map<String, String> filter =
+          ImmutableMap.of(RangerRef.SEARCH_FILTER_SERVICE_NAME, RANGER_HIVE_REPO_NAME);
+      List<RangerService> services = rangerClient.findServices(filter);
+      Assertions.assertEquals(services.get(0).getType(), RANGER_HIVE_TYPE);
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java:131]: Put the expected value in the first of parameters is preferable.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:2]: Can the the license pass CI check?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.integration.test.authorization.ranger;
+
+import org.apache.ranger.plugin.util.SearchFilter;
+
+public class RangerRef {
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:9]: What's the meaning of `ref` here?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  public static final String RESOURCE_DATABASE = "database"; // Hive resource database name
+  public static final String RESOURCE_TABLE = "table"; // Hive resource table name
+  public static final String RESOURCE_COLUMN = "column"; // Hive resource column name
+  public static final String RESOURCE_PATH = "path"; // HDFS resource path name
+  public static final String SEARCH_FILTER_DATABASE =
+      SearchFilter.RESOURCE_PREFIX + RESOURCE_DATABASE;
+  public static final String SEARCH_FILTER_TABLE = SearchFilter.RESOURCE_PREFIX + RESOURCE_TABLE;
+  public static final String SEARCH_FILTER_COLUMN = SearchFilter.RESOURCE_PREFIX + RESOURCE_COLUMN;
+  public static final String SEARCH_FILTER_PATH = SearchFilter.RESOURCE_PREFIX + RESOURCE_PATH;
+  public static final String SERVICE_TYPE_HFDS = "hdfs"; // HDFS service type
```
Comment (yuqi1129) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:26]: SERVICE_TYPE_HFDS typo

Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:26]: DONE.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java (hunk 9 above)
Comment (justinmclean) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:2]: No it will not we need to use the ASF license now

File: dev/docker/hive/Dockerfile [hunk 12]
```diff
@@ -83,15 +86,31 @@ RUN echo "HADOOP_CONF_DIR=${HADOOP_CONF_DIR}" >> /etc/environment
 RUN echo "HADOOP_CLASSPATH=${JAVA_HOME}/lib/tools.jar" >> /etc/environment
 RUN echo "YARN_HOME=${YARN_HOME}" >> /etc/environment
 RUN echo "HIVE_HOME=${HIVE_HOME}" >> /etc/environment
+RUN echo "ZK_HOME=${ZK_HOME}" >> /etc/environment
 RUN echo "PATH=${PATH}" >> /etc/environment
 RUN echo "CLASSPATH=${CLASSPATH}" >> /etc/environment
 RUN echo "LD_LIBRARY_PATH=${LD_LIBRARY_PATH}" >> /etc/environment
 
+################################################################################
+# install zoofkeeper
```
[the remark this answers is not in the document]
Comment (xunliu) [reply dev/docker/hive/Dockerfile:95]: FIXED

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.ranger.plugin.model.RangerPolicy;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+@Tag("gravitino-docker-it")
```
[the remark this answers is not in the document]
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java:27]: Fixed.

File: dev/docker/hive/Dockerfile (hunk 4 above)
Comment (xunliu) [reply dev/docker/hive/Dockerfile:109]: Moved this line to `startup.sh` file.

File: integration-test/build.gradle.kts (hunk 5 above)
Comment (xunliu) [reply integration-test/build.gradle.kts:112]: Removed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java (hunk 6 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerHiveIT.java:33]: Fixed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java (hunk 7 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java:205]: Removed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java (hunk 8 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerIT.java:131]: Fixed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java (hunk 10 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java:9]: Renamed to `RangerDefines`

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java (hunk 9 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerRef.java]: Fixed

**Comment by yuqi1129** [comment]
@jerryshao @mchades , Please help to review it, I'm not quite familiar with Ranger and can't help too much with Ranger-related code.

File: catalogs/catalog-hadoop/src/test/java/com/datastrato/gravitino/catalog/hadoop/integration/test/HadoopCatalogIT.java [hunk 14]
```diff
@@ -68,7 +68,7 @@ public class HadoopCatalogIT extends AbstractIT {
 
   @BeforeAll
   public static void setup() throws IOException {
-    containerSuite.startHiveContainer();
+    containerSuite.startHiveContainer(false, ImmutableMap.of());
```
Comment (yuqi1129) [inline catalogs/catalog-hadoop/src/test/java/com/datastrato/gravitino/catalog/hadoop/integration/test/HadoopCatalogIT.java:71]: I suggest you can add a extra method in container suite like 

```java
startHiveContainer(boolean xxx) {
  startHiveContainer(xxx, ImmutableMap.of()
}

startHiveContainer(boolean xxx, Map<> xxx) {
    // core logic
}
```

In this way, we can avoid some unnecessary changes.

File: dev/docker/hive/Dockerfile [hunk 15]
```diff
@@ -22,7 +22,8 @@ LABEL maintainer="support@datastrato.com"
 
 ARG HADOOP_VERSION
 ARG HIVE_VERSION
-ARG JDBC_DIVER_PACKAGE_NAME
```
Comment (mchades) [inline dev/docker/hive/Dockerfile:25]: why remove this?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerDefines.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
```
Comment (mchades) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerDefines.java:2]: The license need to be changed

**Review by mchades — COMMENTED** [review]
please check the license of all new files

**Comment by yuqi1129** [comment]
@unknowntpo @xunliu Please resolve the comments and conflicts.

File: catalogs/catalog-hadoop/src/test/java/com/datastrato/gravitino/catalog/hadoop/integration/test/HadoopCatalogIT.java (hunk 14 above)
Comment (xunliu) [reply catalogs/catalog-hadoop/src/test/java/com/datastrato/gravitino/catalog/hadoop/integration/test/HadoopCatalogIT.java]: Removed these code.

File: dev/docker/hive/Dockerfile (hunk 15 above)
Comment (xunliu) [reply dev/docker/hive/Dockerfile:25]: Currently, The `JDBC_DIVER_PACKAGE_NAME` argument unused, So we can delete it.

File: docs/publish-docker-images.md [hunk 17]
```diff
@@ -25,8 +25,9 @@ You can use GitHub actions to publish Docker images to the Docker Hub repository
    + `datastrato/gravitino-ci-trino`.
    + Future plans include support for other data sources.
 5. Input the `tag name`, for example: `0.1.0`, Then build and push the Docker image name as `datastrato/{image-name}:0.1.0`.
-6. You must enter the correct `publish docker token` before you can execute run `Publish Docker Image` workflow.
-7. Wait for the workflow to complete. You can see a new Docker image shown in the [Datastrato Docker Hub](https://hub.docker.com/u/datastrato) repository.
+6. Input the `envoriments`, for example: `HADOOP_VERSION=3.1.0 HIVE_VERSION=3.1.3`, Then use special version to install HDFS and Hive.
```
[the remark this answers is not in the document]
Comment (xunliu) [reply docs/publish-docker-images.md]: DONE.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerDefines.java (hunk 16 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/authorization/ranger/RangerDefines.java]: Done.

**Review by xunliu — COMMENTED** [review]
@mchades @yuqi1129 I fixed all issues, Please help me review again, Thanks!

File: docs/docker-image-details.md [hunk 18]
```diff
@@ -118,6 +118,21 @@ You can use this kind of image to test the catalog of Apache Hive.
 
 Changelog
 
+- gravitino-ci-hive:0.1.13
+  - Enable Ranger plugin in the Hive and HDFS
+  - Hive version 3.1.3
+  - HDFS version 3.1.0
```
Comment (mchades) [inline docs/docker-image-details.md:124]: Have you upgraded the versions of Hive and Hadoop from 2.x to 3.x?

All implementations and tests are based on version 2.x. I am concerned that upgrading to version 3.x may introduce potential compatibility issues that have not been tested.

File: integration-test/build.gradle.kts [hunk 19]
```diff
@@ -122,6 +122,9 @@ dependencies {
   testImplementation(libs.trino.client) {
     exclude("jakarta.annotation")
   }
+  testImplementation(libs.hive2.jdbc) {
```
Comment (yuqi1129) [inline integration-test/build.gradle.kts:125]: Please sort them alphabetically

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/HiveContainer.java [hunk 20]
```diff
@@ -201,22 +205,34 @@ protected boolean checkContainerStatus(int retryLimit) {
   }
 
   public static class Builder extends BaseContainer.Builder<Builder, HiveContainer> {
+    boolean rangerEnablePlugin = false;
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/HiveContainer.java:208]: enableRangerPlugin?

**Comment by unknowntpo** [comment]
This commit
[WIP: support hive2,hive3,hadoop2,hadoop3 in one image](https://github.com/apache/gravitino/pull/3912/commits/3dbce4371cc57b2f384a60b8535c930d98e74b9e)

will be reverted, just for testing.

File: dev/docker/hive/Dockerfile [hunk 21]
```diff
@@ -139,9 +147,15 @@ ADD mapred-site.xml ${HADOOP_CONF_DIR}/mapred-site.xml
 ADD check-status.sh /tmp/check-status.sh
 
 ################################################################################
-# install hive
-ADD packages/apache-hive-${HIVE_VERSION}-bin.tar.gz /opt/
-RUN ln -s /opt/apache-hive-${HIVE_VERSION}-bin ${HIVE_HOME}
+RUN mkdir -p ${HIVE_HOME}/lib
```
Comment (xunliu) [inline dev/docker/hive/Dockerfile:150]: I think we didn't need to prepare to create `${HIVE_HOME}/lib` directory.

We only move `RUN ln -s /opt/mysql-connector-java-${MYSQL_JDBC_DRIVER_VERSION}/* ${HIVE_HOME}/lib` to `startup.sh`

File: dev/docker/hive/Dockerfile [hunk 22]
```diff
@@ -181,6 +195,9 @@ RUN chown -R datastrato:hadoop /home/datastrato
 RUN rm -rf /tmp/packages
 RUN rm -rf /var/lib/apt/lists/*
 
+RUN touch ${HIVE_HOME}/conf/orig.txt
+RUN touch ${HIVE_HOME}/lib/orig.txt
```
Comment (xunliu) [inline dev/docker/hive/Dockerfile:199]: Why do we need to add these codes? is a temporary test code?

File: dev/docker/hive/start.sh [hunk 23]
```diff
@@ -18,6 +18,18 @@
 # under the License.
 #
 
+echo "before link"
+ls -la ${HIVE_HOME}
+
+if [[ "${ENABLE_RANGER_PLUGIN}" == "true" ]]; then
```
Comment (xunliu) [inline dev/docker/hive/start.sh:24]: I think we need a new environment `HIVE_VERSION`.

```
if [[ "${HIVE_VERSION}" == "hive3" ]]; then
```

File: dev/docker/hive/Dockerfile (hunk 22 above)
Comment (unknowntpo) [reply dev/docker/hive/Dockerfile]: Sorry, removed.

File: dev/docker/hive/Dockerfile (hunk 21 above)
Comment (unknowntpo) [reply dev/docker/hive/Dockerfile]: done.

**Comment by unknowntpo** [comment]
I've publish new hive image to my docker hub repo for testing, to use this image, please set this line to `integration-test/build.gradle.kts`:

```
      environment("GRAVITINO_CI_RANGER_HIVE_DOCKER_IMAGE", "unknowntpo/gravitino-ci-hive:ranger-plugin") // Support Ranger plugin
```

File: dev/docker/build-docker.sh [hunk 24]
```diff
@@ -122,14 +134,18 @@ fi
 cd ${script_dir}/${component_type}
 if [[ "${platform_type}" == "all" ]]; then
   if [ ${build_latest} -eq 1 ]; then
-    docker buildx build --no-cache --pull --platform=linux/amd64,linux/arm64 ${build_args} --push --progress plain -f Dockerfile -t ${image_name}:latest -t ${image_name}:${tag_name} .
+    # FIXME: need to add back --no-cache
```
Comment (xunliu) [inline dev/docker/build-docker.sh:137]: @unknowntpo Do we need to roll back the changed `--no-cache` in this PR?

File: dev/docker/hive/Dockerfile [hunk 25]
```diff
@@ -97,54 +122,54 @@ RUN echo "HADOOP_CONF_DIR=${HADOOP_CONF_DIR}" >> /etc/environment
...
-
-ADD core-site.xml ${HADOOP_CONF_DIR}/core-site.xml
-ADD hadoop-env.sh ${HADOOP_CONF_DIR}/hadoop-env.sh
-ADD hdfs-site.xml ${HADOOP_CONF_DIR}/hdfs-site.xml
-ADD mapred-site.xml ${HADOOP_CONF_DIR}/mapred-site.xml
-ADD check-status.sh /tmp/check-status.sh
+# install zookeeper
+ADD packages/zookeeper-${ZOOKEEPER_VERSION}.tar.gz /opt/
+RUN ln -s /opt/zookeeper-${ZOOKEEPER_VERSION} ${ZK_HOME}
+RUN mkdir $ZK_HOME/data
```
Comment (xunliu) [inline dev/docker/hive/Dockerfile:134]: @unknowntpo Please change to `${ZK_HOME}`

File: docs/publish-docker-images.md [hunk 26]
```diff
@@ -25,8 +25,9 @@ You can use GitHub actions to publish Docker images to the Docker Hub repository
    + `datastrato/gravitino-ci-trino`.
    + Future plans include support for other data sources.
 5. Input the `tag name`, for example: `0.1.0`, Then build and push the Docker image name as `datastrato/{image-name}:0.1.0`.
-6. You must enter the correct `publish docker token` before you can execute run `Publish Docker Image` workflow.
-7. Wait for the workflow to complete. You can see a new Docker image shown in the [Datastrato Docker Hub](https://hub.docker.com/u/datastrato) repository.
+6. Input the `environment`, for example: `HADOOP_VERSION=3.1.0 HIVE_VERSION=3.1.3`, Then use special version to install HDFS and Hive.
+7. You must enter the correct `publish docker token` before you can execute run `Publish Docker Image` workflow.
+8. Wait for the workflow to complete. You can see a new Docker image shown in the [Datastrato Docker Hub](https://hub.docker.com/u/datastrato) repository.
```
Comment (xunliu) [inline docs/publish-docker-images.md:30]: @unknowntpo Please roll back these changed.

File: dev/docker/hive/Dockerfile (hunk 25 above)
Comment (unknowntpo) [reply dev/docker/hive/Dockerfile]: done

File: docs/publish-docker-images.md (hunk 26 above)
Comment (unknowntpo) [reply docs/publish-docker-images.md]: done

File: dev/docker/build-docker.sh (hunk 24 above)
Comment (unknowntpo) [reply dev/docker/build-docker.sh]: rollbacked.

File: dev/docker/build-docker.sh [hunk 27]
```diff
@@ -87,7 +87,19 @@ fi
...
+  --build-arg HADOOP_PACKAGE_NAME=${HADOOP_PACKAGE_NAME} \
+  --build-arg HIVE_PACKAGE_NAME=${HIVE_PACKAGE_NAME} \
+  --build-arg HADOOP_VERSION=${HADOOP_VERSION} \
+  --build-arg HIVE_VERSION=${HIVE_VERSION} \
+  --build-arg MYSQL_JDBC_DRIVER_VERSION=${MYSQL_JDBC_DRIVER_VERSION} \
+  --build-arg RANGER_VERSION=${RANGER_VERSION} \
+  --build-arg ZOOKEEPER_VERSION=${ZOOKEEPER_VERSION} \
+  --build-arg HIVE2_VERSION=${HIVE2_VERSION} \
+  --build-arg HIVE3_VERSION=${HIVE3_VERSION}
+  --build-arg HADOOP2_VERSION=${HADOOP2_VERSION}
```
Comment (xunliu) [inline dev/docker/build-docker.sh:100]: I think need to add `\` at the end of the line.

File: docs/publish-docker-images.md [hunk 28]
```diff
@@ -25,8 +25,6 @@ You can use GitHub actions to publish Docker images to the Docker Hub repository
    + `datastrato/gravitino-ci-trino`.
    + Future plans include support for other data sources.
 5. Input the `tag name`, for example: `0.1.0`, Then build and push the Docker image name as `datastrato/{image-name}:0.1.0`.
-6. You must enter the correct `publish docker token` before you can execute run `Publish Docker Image` workflow.
-7. Wait for the workflow to complete. You can see a new Docker image shown in the [Datastrato Docker Hub](https://hub.docker.com/u/datastrato) repository.
```
Comment (xunliu) [inline docs/publish-docker-images.md:29]: We didn't delete these two lines.

Comment (xunliu) [reply docs/publish-docker-images.md]: @unknowntpo I help you fixed this issue.

File: dev/docker/build-docker.sh (hunk 27 above)
Comment (xunliu) [reply dev/docker/build-docker.sh]: @unknowntpo I help you fixed this issue.

**Review by xunliu — APPROVED** [review]
@unknowntpo Thank you for your contributions.
LGTM

**Review by yuqi1129 — APPROVED** [review]
LGTM
